README.md
# react-animated-text-content

[![npm](https://img.shields.io/npm/dt/react-animated-text-content.svg?style=flat-square)](https://www.npmjs.com/package/react-animated-text-content)
[![npm](https://img.shields.io/npm/v/react-animated-text-content.svg?style=flat-square)](https://www.npmjs.com/package/react-animated-text-content)

A component to animate your text in React! Use predefined animation type or compose your own one.

[Demo](https://kalkson.github.io/react-animated-text-content-demo/)

## Install

```bash
  npm install --save react-animated-text-content
```

## Note

Component has animate on scroll functionality built in (whith `Intersection Observer` api) so you don't have to use 3rd party libs to achieve this. If you want to find out more about Intersection Observer check mdn page about that topic [here](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API).

### Important

Be careful with using this library and don't overdo with it. To achieve text animation like this way, the big chunck of HTML must be rendered and every text fragment has separated animation properties depending on props passed. This could cause serious performance leaks (especially in case of `chars` type) but it may not. Just use it with cool head and don't use long texts to animate!

## Using

```jsx
import AnimatedText from 'react-animated-text-content';

<AnimatedText
  type="words" // animate words or chars
  animation={{
    x: '200px',
    y: '-20px',
    scale: 1.1,
    ease: 'ease-in-out',
  }}
  animationType="float"
  interval={0.06}
  duration={0.8}
  tag="p"
  className="animated-paragraph"
  includeWhiteSpaces
  threshold={0.1}
  rootMargin="20%"
>
  Content to animate
</AnimatedText>;
```

The `AnimatedText` component render way is fully customizable. If you want, you can use predefined animation or compose your own one. In order to check and test predefined animation types,
visit [demo](https://kalkson.github.io/react-animated-text-content-demo/) site. You can use animation generator there as well and generate code to paste in.

- define own animation object and pass it as props,
- if you use `animationType` prop, the `animation` prop is not valid anymore (no effect),
- you can oberwrite default `duration` and `interval` values (from `animationType` as well) by defining pure props. These props always has higher priority. If no specified, the fallback with default value is used

## props

All props are optional.

| name                   | type                                                | default          | description                                                                                                                                                                                                                                               |
| ---------------------- | --------------------------------------------------- | ---------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **type**               | 'words'                                             | `'chars'`        | specify animation render way. Depending on `type` the HTML of component is drastically different                                                                                                                                                          |
| **animation**          | {x: string, y: string, scale: number, ease: string} | `{ y: '-30px' }` | define animation from state with optional properties like: `x`, `y`, `ease`, `scale`                                                                                                                                                                      |
| **animationType**      | string                                              | -                | use predefined animation (check currently available ones at demo page)                                                                                                                                                                                    |
| **children**           | ReactText                                           | ''               | the most important prop. Currently as ReactText type so no addiditional elements are allowed like `<br />`. If so, the error will be catched                                                                                                              |
| **interval**           | number                                              | `0.04`           | pretty same as `animation` prop object property. Use this version of props to overwrite default value from `animationType` or `animation` object. `s` (seconds) unit. Says to the component how quickly next fragments of the text should be animated     |
| **duration**           | number                                              | `0.4`            | pretty same as `animation` prop object property. Use this version of props to overwrite default value from `animationType` or `animation` object. `s` (seconds) unit. Says to the component how long will animation be performing in case of one fragment |
| **tag**                | string                                              | `'div'`          | specify custom HTML tag you want to render as you animated text parent. Be aware of that some of tags could break animation behavior or event stop runtime so use it carefully                                                                            |
| **className**          | string                                              | -                | define own class for component                                                                                                                                                                                                                            |
| **includeWhiteSpaces** | boolean                                             | `false`          | white spaces could be rendered in separated tags or as blank chars. Change it only if you must. When set to `true`, performance goes down                                                                                                                 |
| **threshold**          | number                                              | `0`              | value for Intersection Observer. Find out more here [docs](https://developer.mozilla.org/en-us/docs/web/api/intersection_observer_api). Otherwise, leave it default                                                                                       |
| **rootMargin**         | string                                              | `0`              | value for Intersection Observer. Find out more here [docs](https://developer.mozilla.org/en-us/docs/web/api/intersection_observer_api). Otherwise, leave it default. `px` and `%` units are allowed                                                       |
| **{...props}**         | -                                                   | -                | all props like `style`, `aria-`, `id` are allowed                                                                                                                                                                                                         |

## LICENSE

[MIT](LICENSE)


CHANGELOG.md
## 1.0.1

- fixed types bundling


## 1.0.0

- created initial version


src/tests/helpers.test.ts
import { concatFragments, splitChildrenToFragments } from '../helpers';

const MOCKED_TEXT = 'Some dummy content to animate';

describe('splitChildrenToFragments', () => {
  test('should return proper result', () => {
    const expectedResult = ['Some', 'dummy', 'content', 'to', 'animate'];

    expect(splitChildrenToFragments(MOCKED_TEXT)).toEqual(expectedResult);
  });
});

describe('concatFragment', () => {
  test('should return proper result', () => {
    const result = concatFragments(MOCKED_TEXT);

    expect(result[0]).toEqual('Some');
    expect(result[result.length - 1]).toEqual('animate');

    result.forEach((fragment, index) => {
      if (index % 2 === 1) {
        expect(fragment).toEqual(' ');
      }
    });
  });

  test('should return proper result when type is number', () => {
    const result = concatFragments(2137);

    expect(result[0]).toEqual('2137');
  });
});


rollup.config.js
import sass from 'rollup-plugin-sass';
import typescript from 'rollup-plugin-typescript2';
import { terser } from 'rollup-plugin-terser';
import externalDeps from 'rollup-plugin-peer-deps-external';
import replace from '@rollup/plugin-replace';

// import pkg from './package.json';

const external = ['react', 'react-dom'];
const globals = {
  react: 'React',
  uuid: 'uuid',
  'styled-components': 'styled',
};

export default [
  {
    input: 'src/index.tsx',
    output: [
      {
        name: 'AnimatedText',
        file: 'dist/react-animated-text-content.development.js',
        format: 'umd',
        exports: 'named',
        sourcemap: true,
        strict: false,
        compact: true,
        globals,
      },
    ],
    plugins: [sass({ insert: true }), typescript()],
    external,
  },
  {
    input: 'src/index.tsx',
    output: [
      {
        name: 'AnimatedText',
        file: 'dist/react-animated-text-content.production.js',
        format: 'umd',
        exports: 'named',
        sourcemap: true,
        strict: true,
        compact: true,
        globals,
      },
    ],
    plugins: [
      sass({ insert: true }),
      typescript(),
      replace({ 'process.env.NODE_ENV': `"production"`, delimiters: ['', ''] }),
      externalDeps(),
      terser(),
    ],
    external: ['react', 'react-dom'],
  },
];


src/styles/AnimatedFragment.styled.ts
import styled, { css } from 'styled-components';
import { AnimationShapeType } from '..';

export const AnimatedFragment = styled.span<{
  count?: number;
  interval?: number;
}>`
  position: relative;
  display: inline-block;
  visibility: hidden;
`;

export const StyledWrapper = styled.div<{
  uid: string;
  count: number;
  interval: number;
  duration: number;
  shouldAnimate: boolean;
  animation: AnimationShapeType;
}>`
  @keyframes fragmentletter-${({ uid }) => uid} {
    0% {
      opacity: 0;
      visibility: hidden;
      transform: translateX(${({ animation: { x } }) => x ?? 0})
        translateY(${({ animation: { y } }) => y ?? 0})
        scale(${({ animation: { scale } }) => scale ?? 1});
    }
    100% {
      oapcity: 1;
      visibility: visible;
      transform: translateX(0) translateY(0) scale(1);
    }
  }

  & > span {
    display: inline-block;
  }

  & ${AnimatedFragment} {
    text-decoration: inherit;
    vertical-align: inherit;

    animation: ${({ animation: { ease }, duration }) =>
      `${duration}s ${ease ?? 'ease-in-out'} forwards;`};

    animation-name: ${({ shouldAnimate, uid }) =>
      shouldAnimate ? `fragmentletter-${uid}` : ''};
  }

  ${({ count = 0, interval }) => {
    let styles = '';

    for (let i = 0; i < count; i += 1) {
      styles += `${AnimatedFragment}:nth-of-type(${i + 1}) {
            animation-delay: ${(interval || 0.04) * i}s;
          }`;
    }

    return css`
      ${styles}
    `;
  }}
`;


src/constants.ts
import { AnimationType, ExtendedAnimationShapeType } from './types';

export const PREDEFINED_ANIMATIONS: Record<
  AnimationType,
  ExtendedAnimationShapeType
> = {
  blocks: {
    y: '60px',
    interval: 0.07,
    duration: 0.7,
  },
  wave: {
    y: '40px',
    interval: 0.04,
    duration: 0.4,
    ease: 'ease-in-out',
  },
  float: {
    x: '200px',
    ease: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
    duration: 1.2,
    interval: 0.1,
  },
  bounce: { y: '200px', ease: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)' },
  throw: {
    y: '200px',
    scale: 2,
    interval: 0.07,
  },
  diagonal: {
    x: '200px',
    y: '-200px',
    ease: 'cubic-bezier(0.68, -4.55, 0.265, 1.55)',
  },
  rifle: {
    y: '100px',
    x: '-500px',
    ease: 'ease-in',
    duration: 0.3,
    interval: 0.03,
  },
  lights: {
    y: '-20px',
    ease: 'ease-out',
    duration: 1,
    interval: 0.4,
  },
};

export const WHITE_SPACE_CODE = ' ';
export const DEFAULT_INTERVAL = 0.04;
export const DEFAULT_DURATION = 0.4;
export const DEFAULT_TYPE = 'words';
export const DEFAULT_ANIMATION = {
  y: '-30px',
};
export const DEFAULT_ANIMATION_TYPE = null;
export const DEFAULT_TAG = 'div';
export const DEFAULT_INTESECTION_OBSERVER_ROOT_MARGIN = '20%';
export const DEFAULT_THRESHOLD = 0;
